feat(overlay): add keyboard dispatcher for targeting correct overlay

willshowell opened the pull request:
Fixes https://github.com/angular/material2/issues/6330
Fixes https://github.com/angular/material2/issues/3251
Closes https://github.com/angular/material2/pull/3460

jelbourn commented on src/cdk/overlay/overlay-ref.ts:
```diff
@@ -140,6 +145,16 @@ export class OverlayRef implements PortalHost {
     return this._detachments.asObservable();
   }
 
+  /** Returns an observable that emits when the overlay has been disposed. */
```
nit: prefer "Gets" to "Returns"

jelbourn commented on src/cdk/overlay/overlay.ts:
```diff
@@ -87,4 +93,18 @@ export class Overlay {
+   * Subscribe to attach and detach events and register them accordingly
+   * with the keyboard dispatcher service.
+   */
+  private _trackOverlayAttachments(overlayRef: OverlayRef): void {
```
Could each `OverlayRef` to this itself in its constructor?

willshowell commented:
@jelbourn comments addressed. Also added a demo to the overlay.

crisbeto commented on src/lib/dialog/dialog.ts:
```diff
@@ -241,6 +236,15 @@ export class MdDialog {
+    RxChain.from(overlayRef.keyboardEvents())
+      .call(filter, event => event.type === 'keydown' && event.keyCode === ESCAPE)
+      .subscribe(() => {
+        if (!dialogRef.disableClose) {
```
Might be a bit more concise to have this in `filter`, instead of checking it in the `subscribe`.

crisbeto commented on src/cdk/overlay/keyboard/overlay-keyboard-dispatcher.ts:
```diff
@@ -0,0 +1,98 @@
+    const bodyKeyboardEvents = merge(
+        fromEvent(document.body, 'keydown'),
+        fromEvent(document.body, 'keypress'),
+        fromEvent(document.body, 'keyup')) as Observable<KeyboardEvent>;
```
Instead of casting it to an `Observable<KeyboardEvent>`, you can pass the `KeyboardEvent` as a generic: `merge<KeyboardEvent>`.

crisbeto commented on src/cdk/overlay/keyboard/overlay-keyboard-dispatcher.ts:
```diff
@@ -0,0 +1,98 @@
+  private _dispatchKeyboardEvents(): void {
+    const bodyKeyboardEvents = merge(
+        fromEvent(document.body, 'keydown'),
+        fromEvent(document.body, 'keypress'),
```
I'm not sure that this should handle `keydown`, `keypress` and `keyup` @jelbourn. Since for 99% of cases you'd only need `keydown`, it means that every time you consume the service, you'd have to have a `filter(event => event.type === 'keydown')`.

willshowell replied:
I'd be happy with this. I was aiming for generic, but the api could be "rebranded" 'to just `keydownEvents`.

crisbeto commented on src/cdk/overlay/keyboard/overlay-keyboard-dispatcher.ts:
```diff
@@ -0,0 +1,98 @@
+  /** Select the appropriate overlay from a list of overlays and a keyboard event. */
+  private _selectOverlayFromEvent(overlays: OverlayRef[], event: KeyboardEvent): OverlayRef {
+    // Check if any overlays contain the event
+    const targetedOverlay = this._attachedOverlays.find(overlay =>
```
This doesn't handle if the keyboard event came directly from the overlay element. You can tweak the predicate to include `overlay.overlayElement === event.target`.

crisbeto commented on src/cdk/overlay/keyboard/overlay-keyboard-dispatcher.ts:
```diff
@@ -0,0 +1,98 @@
+  }
+
+  /** Select the appropriate overlay from a list of overlays and a keyboard event. */
+  private _selectOverlayFromEvent(overlays: OverlayRef[], event: KeyboardEvent): OverlayRef {
```
Since the `overlays` is always going to be `this._attachedOverlays`, you can take it from the class property instead of passing it in.

crisbeto commented on src/cdk/overlay/overlay-ref.ts:
```diff
@@ -22,12 +23,16 @@ export class OverlayRef implements PortalHost {
   private _detachments = new Subject<void>();
+  private _disposed = new Subject<void>();
+
+  _dispatchedKeyboardEvents = new Subject<KeyboardEvent>();
```
Instead of having the `OverlayRef` add and remove itself from the dispatcher, a cleaner approach would be to only have it expose the stream of keyboard events. Then the `OverlayRef.keyboardEvents`  can be simplified to:

```ts
// note: don't get too caught up in the naming
return this.keyboardDispatcher.events.filter(event => event.target === this.overlayElement);
```

This less room for mistakes when managing the overlays and it also removes the circular dependency between the `OverlayRef` and the dispatcher.

willshowell replied:
Hm I don't entirely follow.

1. I kind of see how I could remove the `Subject`; maybe if I created a new event type that paired overlays to keydowns, and then the overlay just filters out ones directed at itself and maps to the keydown.
2. An important piece of this is dispatching an event to the most recently attached overlay. Often, events land on the body with no other context. That's the whole point of tracking overlays as they are attached and detached.

crisbeto replied:
Alright, I see. I guess that in most cases the top-most overlay would be the one dispatching the keyboard events, but that won't always be the case (like with `md-autocomplete`). Let's keep it.

willshowell commented:
@crisbeto comments addressed. Please take another look!

crisbeto commented on src/cdk/overlay/keyboard/overlay-keyboard-dispatcher.ts:
```diff
@@ -0,0 +1,96 @@
+  private _selectOverlayFromEvent(event: KeyboardEvent): OverlayRef {
+    // Check if any overlays contain the event
+    const targetedOverlay = this._attachedOverlays.find(overlay => {
+      return overlay.overlayElement.contains(event.target as HTMLElement) ||
```
I'd move the `overlay.overlayElement === event.target` check first since it could potentially exit sooner.

crisbeto commented on src/cdk/overlay/keyboard/overlay-keyboard-dispatcher.ts:
```diff
@@ -0,0 +1,96 @@
+    });
+
+    // Use that overlay if it exists, otherwise choose the most recently attached one
+    return targetedOverlay ?
```
`return targetedOverlay || this._attachedOverlays[this._attachedOverlays.length - 1]`.

crisbeto commented on src/cdk/overlay/overlay-ref.ts:
```diff
@@ -23,11 +24,14 @@ export class OverlayRef implements PortalHost {
   private _attachments = new Subject<void>();
   private _detachments = new Subject<void>();
 
+  _dispatchedKeydownEvents = new Subject<KeyboardEvent>();
```
Add a description to this one. Also potentially rename it to something like `_keydownEvents`.

crisbeto commented:
@willshowell looks like the `prerender` task is still failing. You can debug it locally using `gulp ci:prerender`. Ping me to mark as merge ready.

willshowell commented:
@crisbeto I had issues with `fromEvent` not working well with the `DOCUMENT` token, so I opted to make the dispatch start lazily. PTAL.

crisbeto commented on src/cdk/overlay/keyboard/overlay-keyboard-dispatcher.ts:
```diff
@@ -0,0 +1,98 @@
+   * Subscribe to keydown events that land on the body and dispatch those
+   * events to the appropriate overlay.
+   */
+  private _dispatchKeydownEvents(): void {
```
Judging by the name of the method, I'd assume that it would dispatch the events immediately, but that's not what it does. Something along the lines of `_subscribeToKeydownEvents` might be more appropriate.

willshowell commented:
Rebased. cc @kara

willshowell commented:
@kara is there anything I need to do regarding the presubmit failure?

andrewseguin commented:
Getting some AOT issue:

```
Error: Error encountered resolving symbol values statically. Only initialized variables and constants can be referenced because the value of this variable is needed by the template compiler (position 10:22 in the original .ts file), resolving symbol DOCUMENT in /platform-browser/src/dom/dom_tokens.d.ts, resolving symbol DOCUMENT in /platform-browser/src/platform-browser.d.ts, resolving symbol DOCUMENT in /platform-browser/public_api.d.ts, resolving symbol DOCUMENT in /platform-browser/index.d.ts, resolving symbol DOCUMENT in platform_browser_safe/index.d.ts, resolving symbol OVERLAY_KEYBOARD_DISPATCHER_PROVIDER in /src/cdk/overlay/keyboard/overlay-keyboard-dispatcher.ts, resolving symbol OVERLAY_PROVIDERS in /src/cdk/overlay/overlay-module.ts, resolving symbol OverlayModule in /src/cdk/overlay/overlay-module.ts, resolving symbol OverlayModule in /src/cdk/overlay/overlay-module.ts

```

Can you try building in AOT?

willshowell commented:
@andrewseguin I wasn't able to reproduce, but I've rebased and updated the license.

I'm passing on `gulp ci:lint, aot, prerender` (`ci:test` is broken from something else).

mmalerba commented:
Is injecting `DOCUMENT` instead of using `document` critical here? Removing the `DOCUMENT` stuff fixes the error @andrewseguin  mentioned

jelbourn commented:
@mmalerba we should only need to use `DOCUMENT` if we're mocking something in a test

mmalerba commented:
I think `DOCUMENT` is special somehow, I tried creating a different injection token, `OVERLAY_DOCUMENT`, and everything worked fine. So we should switch it and if we need it for testing just create our own provider. @willshowell Once that's done I can try again on presubmitting.

jelbourn commented:
Oh yeah, now I remember. I had to do the same thing for the tests for `Directionality`. `BidiModule` has
```ts
{provide: DIR_DOCUMENT, useExisting: DOCUMENT},
```

willshowell commented:
Doesn't seem there's any need to mock `document` for the tests, so I've removed the token.

jelbourn commented:
@mmalerba should be ready for another spin
